### PR TITLE
[mir-inlining] Don't inline virtual calls

### PR DESCRIPTION
**Pull request by wesleywiser**

### CODE REVIEW

File: src/librustc_mir/transform/inline.rs [hunk 1]
```diff
@@ -100,12 +100,19 @@ impl<'a, 'tcx> Inliner<'a, 'tcx> {
...
-                                callsites.push_back(CallSite {
-                                    callee: instance.def_id(),
-                                    substs: instance.substs,
-                                    bb,
-                                    location: terminator.source_info
-                                });
+                                let is_virtual =
+                                    if let InstanceDef::Virtual(..) = instance.def {
+                                        true
+                                    } else { false };
```
Comment (varkor) [inline src/librustc_mir/transform/inline.rs]: Nit: `false` should be on its own line.

File: src/test/mir-opt/inline-trait-method.rs [hunk 2]
```diff
@@ -0,0 +1,41 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
```
Comment (varkor) [inline src/test/mir-opt/inline-trait-method.rs]: Copyright is no longer needed.

File: src/test/mir-opt/inline-trait-method.rs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// compile-flags: -Z span_free_formats
+
+fn main() {
+    println!("{}", test(&()));
+}
+
+fn test(x: &X) -> u32 {
```
Comment (varkor) [inline src/test/mir-opt/inline-trait-method.rs]: I think we're moving towards `&dyn X`.

**Comment by varkor** [comment]
Was it inlining the default function implementation, or the implementation for `()` previously? Either way, r=me after the comments are addressed.

**Comment by wesleywiser** [comment]
It was inlining the default implementation so prior to the patch, the test would output `1` instead of `2` like it should.

**Comment by wesleywiser** [comment]
Feedback resolved. Thanks @varkor!

**Comment by RalfJung** [comment]
I think this fixed https://github.com/rust-lang/rust/issues/50411, and running the miri test suite with MIR optimizations. :)

Would be great if the reproducer from https://github.com/rust-lang/rust/issues/50411 could be added as a test case.

**Comment by RalfJung** [comment]
Just, something strange is going on, and while the tests pass locally they fail on CI?!?  See https://github.com/rust-lang/rust/pull/55086.

**Comment by wesleywiser** [comment]
@RalfJung Is there anything that needs to be done related to this bug fix? I see you closed the other pull request.

**Comment by RalfJung** [comment]
Yeah, turns out I was wrong when I thought that the ICE got fixed. :(  Sorry for the noise.

**Comment by wesleywiser** [comment]
No worries :)

**Pull request merged**